### ISSUE TITLE
Why is input_scale needed in TrainableNonLinearity?

### ISSUE THREAD

**Issue by vvolhejn** [ben-hayes/neural-waveshaping-synthesis#9]
Hi, in class `TrainableNonLinearity` there is a parameter [`input_scale`](https://github.com/ben-hayes/neural-waveshaping-synthesis/blob/4a80867957a86c21706a38c2a9d51455b5129ffb/neural_waveshaping_synthesis/models/modules/shaping.py#L21) with which the input is multiplied. However, the class is only used in the `NEWT` class and [when it is used](https://github.com/ben-hayes/neural-waveshaping-synthesis/blob/4a80867957a86c21706a38c2a9d51455b5129ffb/neural_waveshaping_synthesis/models/modules/shaping.py#L74) the input is already scaled by a trainable parameter (the `FiLM` class). Doesn't that make `input_scale` redundant?

Also, I don't think this is mentioned in the paper.

**Comment by ben-hayes** [comment]
Hi @vvolhejn ,

Thanks for the comment, and for clearly taking the time to read the code! Great to know that people are paying attention :)

You're absolutely right in general that consecutive linear transformations like this are redundant — in the context of a neural network with nondecreasing nonlinearities (such as ReLU) this would be a pointless thing to do. When using periodic nonlinearities, however, a scaling parameter is often used to bias the bandwidth of frequencies a given network can produce — see, for example the ![\omega_0](https://latex.codecogs.com/png.image?\dpi{110}%20\omega_0) parameter in [SIREN](https://arxiv.org/abs/2006.09661).

Due to the periodicity of the sinusoidal activations, each parameter's loss surface is tiled with local minima, and due to the harmonic distortion properties of sinusoids the region surrounding each minimum (i.e. the locally pseudoconvex region between the adjacent maxima) results in a different possible range of bandwidths, as higher order distortion products pass the threshold of floating point precision. In our case, the scaling parameter is essentially an initialisation trick that ensures we start gradient descent with the activations of our first layer distributed across a few of these minima, allowing the model to learn which are most favourable. This is similar to what is done in [π-GAN](https://github.com/marcoamonteiro/pi-GAN/blob/0800af72b8a9371b2b62fec2ae69c32994bb802f/siren/siren.py#L302) and [PCINR](https://github.com/janzuiderveld/INR-collection/blob/42f0f99f49a25a92f4a8d079dc88cadf4652ec3b/INR_collection/modules.py#L86), where they directly modify the FiLM scaling parameter. In our case, this parameter actually needn't have been trainable. 

Sadly, due to ISMIR's strict page limits, we didn't have space in the paper to address these details of parameter initialisation or the nuances of working with periodic activations. Our focus was on communicating the salient points of the architecture and experiment results. We will actually be addressing this detail and some related issues in a follow-up paper very soon, so watch this space.

Hope this clears things up for now. Let me know if anything's not clear.

Best
Ben

**Comment by vvolhejn** [comment]
Got it, thanks for the in-depth explanation! I think I missed the fact that the FiLM parameters (fancy way to call an affine function by the way :P) are produced by an NLP, so it's not simply two parameters doing the same thing. It makes sense now.
